### PR TITLE
Support Postgres POSIX regex operators close #4317

### LINKED ISSUE #4317: posix regular expression operator
Currently, for regex match operations we have `_similar` and `_nsimilar` operators. These are based on Postgres' SIMILAR operator which is based on SQL standard definition of regular expression.

Postgres also has POSIX regular expression which might be more common in usage: https://www.postgresql.org/docs/current/functions-matching.html#FUNCTIONS-POSIX-REGEXP

We can add these operators to permissions and graphql schema.

### PR DESCRIPTION
**Pull request by christophediprima**
### Description

Add support for following Postgres operators:
- _regex: ~
- _iregex: ~*
- _nregex: !~
- _niregex: !~*

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR.

### Affected components

- [x] Server
- [x] Console
- [x] CLI
- [x] Docs
- [x] Community Content
- [x] Tests

### Related Issues

#4317

#### Catalog upgrade

Does this PR change Hasura Catalog version?
- [x] No

#### Metadata

Does this PR add a new Metadata feature?
- [x] No

#### GraphQL
- [X] New GraphQL schema is being generated:
   - [X] New fields in `String_comparison_exp`

#### Breaking changes

- [X] No Breaking changes

### CODE REVIEW

**Comment by christophediprima** [comment]
This is my first pull request. It is not fully completed yet. Any comment or help to complete it is welcome :)

**Comment by abooij** [comment]
@christophediprima You can indicate that the PR is not yet ready for review by putting it in a "draft" state. You can do it before you open it, or, for this PR which is already open, click "Convert to draft" on the right-hand side under the "Reviewers" section.

**Comment by christophediprima** [comment]
@abooij I think I am done for the code. Could you help me with the PR remaining checks? I am not sure if they are yes or no.

**Comment by christophediprima** [comment]
@abooij how long do you think it could take to be integrated into a new version?

**Comment by abooij** [comment]
Hey @christophediprima! Sorry, I missed your message from last week. Thanks for reminding me, I'll have a look at this PR on Monday.

File: server/src-lib/Hasura/RQL/IR/BoolExp.hs [hunk 1]
```diff
@@ -244,6 +244,11 @@ data OpExpG (b :: BackendType) a
   | ASIMILAR !a -- similar, regex
   | ANSIMILAR !a-- not similar, regex
 
+  | APOSIXCS !a -- match POSIX case sensitive, regex
+  | APOSIXCI !a -- match POSIX case insensitive, regex
+  | ANPOSIXCS !a-- dont match POSIX case sensitive, regex
+  | ANPOSIXCI !a-- dont match POSIX case insensitive, regex
```
Comment (abooij) [inline server/src-lib/Hasura/RQL/IR/BoolExp.hs:250]: We are in the process of reorganizing and generalizing a bunch of code in the RQL modules. However, it's not necessary for this PR to participate in that process. But let's add a comment that documents what would need to be done in the future here.
```suggestion
  -- Now that in the RQL code we've started to take a "trees that grow"
  -- approach (see PR #6003), we may eventually want to move these
  -- recently added constructors, which correspond to newly supported
  -- Postgres operators, to the backend-specific extensions of this type.
  | APOSIXCS !a -- match POSIX case sensitive, regex
  | APOSIXCI !a -- match POSIX case insensitive, regex
  | ANPOSIXCS !a-- dont match POSIX case sensitive, regex
  | ANPOSIXCI !a-- dont match POSIX case insensitive, regex
```

**Review by abooij — COMMENTED** [review]
Thanks for this PR, @christophediprima. The Haskell code has been developed very nicely, following the pattern of the other operators accurately. I have no substantial comments on the code itself. Please add one comment for future reference (since this work is overlapping some other work that's currently taking place, and needs to be coordinated somewhat). And I've attempted to improve some test descriptions somewhat.

@beerose, who could review the console part of this PR?

@marionschleifer, could you review the docs for this?

@tirumaraiselvan, could you please review the changelog?

**Review by beerose — APPROVED** [review]
Thanks for the PR! 
Approving console changes.

**Review by abooij — APPROVED** [review]

**Review by marionschleifer — APPROVED** [review]
Docs 👍

**Comment by christophediprima** [comment]
Hi @tirumaraiselvan I rebased the changelog.

Thanks for the review @abooij, @beerose and @marionschleifer ! :)

**Comment by christophediprima** [comment]
Should I keep resolving conflicts on `CHANGELOG.md` @abooij? Or will @tirumaraiselvan take care of this? 😅

File: CHANGELOG.md [hunk 2]
```diff
@@ -61,6 +61,8 @@ This release contains the [PDV refactor (#4111)](https://github.com/hasura/graph
 - server: Configurable websocket keep-alive interval. Add `--websocket-keepalive` command-line flag and `HASURA_GRAPHQL_WEBSOCKET_KEEPALIVE` env variable (fix #3539)
 - server: validate remote schema queries (fixes #4143)
 - server: introduce optional custom table name in table configuration to track the table according to the custom name. The `set_table_custom_fields` API has been deprecated, A new API `set_table_customization` has been added to set the configuration. (#3811)
+- server: add support for POSIX operators (close #4317) (#6172)
```
Comment (tirumaraiselvan) [inline CHANGELOG.md:64]: can we just say `server: add support for POSIX operators` ?

File: console/src/components/Common/FilterQuery/types.ts [hunk 3]
```diff
@@ -19,7 +19,11 @@ export type Operator =
   | '$ilike'
   | '$nilike'
   | '$similar'
-  | '$nsimilar';
+  | '$nsimilar'
+  | '$posix_cs'
```
Comment (tirumaraiselvan) [inline console/src/components/Common/FilterQuery/types.ts:23]: Why not call these `$regex`, `$iregex` , `$nregex` and `$inregex` respectively?

File: server/tests-py/test_v1_queries.py [hunk 4]
```diff
@@ -168,6 +168,18 @@ def test_city_where_similar(self, hge_ctx):
...
+
+    def test_city_where_not_posix_cs(self, hge_ctx):
+        check_query_f(hge_ctx, self.dir() + '/select_city_where_not_posix_cs.yaml')
+
+    def test_city_where_posix_ci(self, hge_ctx):
+        check_query_f(hge_ctx, self.dir() + '/select_city_where_posix_ci.yaml')
+
+    def test_city_where_not_posix_ci(self, hge_ctx):
+        check_query_f(hge_ctx, self.dir() + '/select_city_where_not_posix_ci.yaml')
+
```
Comment (tirumaraiselvan) [inline server/tests-py/test_v1_queries.py:182]: Few test cases to use one of the operators in select, update, insert and delete permission would be very useful.

**Review by tirumaraiselvan — CHANGES_REQUESTED** [review]
Thanks for the wonderful PR but I think we should use `regex` as the base operator name and also use the existing conventions for "not" and "case insensitivity" for the derived operators: https://github.com/hasura/graphql-engine/pull/6172/files#r530185924

File: server/tests-py/test_v1_queries.py (hunk 4 above)
Comment (tirumaraiselvan) [reply server/tests-py/test_v1_queries.py:182]: Something which checks for valid emails maybe?

File: CHANGELOG.md [hunk 5]
```diff
@@ -69,6 +71,8 @@ This release contains the [PDV refactor (#4111)](https://github.com/hasura/graph
 - console: allow setting computed fields for views (close #6168) (#6174)
 - cli: add missing global flags for seed command (#5565)
 - cli: allow seeds as alias for seed command (#5693)
+- cli-ext: fix get-shared-modules script execution on mac (#6172)
```
Comment (tirumaraiselvan) [inline CHANGELOG.md:74]: Might seem like nitpick but is it possible to make the `cli-ext` fix in a separate PR?

File: console/src/components/Common/FilterQuery/types.ts (hunk 3 above)
Comment (christophediprima) [reply console/src/components/Common/FilterQuery/types.ts:23]: It makes sense, I was not sure about the name as you can also use SQL regex with SIMILAR.
If you are the one that can take the decision I am perfectly OK to change it now. :)

Comment (abooij) [reply console/src/components/Common/FilterQuery/types.ts:23]: I agree that having "regex" in the name is more descriptive than having "posix" in the name. @christophediprima, please make this change :)

(And thanks for being so responsive!)

File: server/tests-py/test_v1_queries.py (hunk 4 above)
Comment (tirumaraiselvan) [reply server/tests-py/test_v1_queries.py:182]: Actually one select permission test is enough.

**Comment by abooij** [comment]
@christophediprima 

> Should I keep resolving conflicts on `CHANGELOG.md` @abooij? Or will @tirumaraiselvan take care of this? sweat_smile

As a rule of thumb, during the review phase of a PR, it's not necessary to resolve every conflict, especially e.g. in the changelog. If there are any substantial conflicts, e.g. another PR has been merged that touches a lot of the same code, then it would be wise to resolve the conflict even before your PR is accepted, so that your PR doesn't go out of date too much.

When you do resolve conflicts, in this repository we prefer that you do so using merge commits rather than rebases and force-pushes, since:

1. we end up squashing the entire PR anyway
2. this way we can see how you resolved certain conflicts, and how you are responding to review comments, etc.

File: server/tests-py/test_v1_queries.py (hunk 4 above)
Comment (christophediprima) [reply server/tests-py/test_v1_queries.py:182]: Could you link me an exemple of such a test with another operator?

Comment (tirumaraiselvan) [reply server/tests-py/test_v1_queries.py:182]: This one has a select permission check for `_in` operator: https://github.com/hasura/graphql-engine/blob/29925eb08de0cbe373f90f6ca6a21d19780a486a/server/tests-py/queries/graphql_query/permissions/setup.yaml#L248

You can perhaps add one in this file itself?

Comment (christophediprima) [reply server/tests-py/test_v1_queries.py:182]: Is that ok? https://github.com/hasura/graphql-engine/pull/6172/commits/d14a419298ce2c977024b9f9c4c5efdc459f52be

**Review by tirumaraiselvan — CHANGES_REQUESTED** [review]
@christophediprima  Thank you. Just the final minor thing: can we revert the changes to `cli-ext/package.json` and `cli-ext/tests/sdl/payload.js` files so they are out of this changeset. I think it's just a lint change.

**Comment by christophediprima** [comment]
@tirumaraiselvan `cli-ext/tests/sdl/payload.js`has some changes. I fixed the linting issue.

**Review by tirumaraiselvan — APPROVED** [review]

**Comment by christophediprima** [comment]
@tirumaraiselvan do I still have to do something or is everything ok? :)

**Comment by abooij** [comment]
@christophediprima Congrats, all lights are green! We'll take this up and let you know if there's anything else you can do.

**Comment by abooij** [comment]
@christophediprima: @hasura-bot seems to have been confused momentarily about your PR, but rest assured it was [merged successfully](https://github.com/hasura/graphql-engine/commit/7b31ff99d16285c4f45bc78e091220686794b1f9)!

**Pull request closed without merging**